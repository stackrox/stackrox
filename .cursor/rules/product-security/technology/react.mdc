---
description: Secure coding rules for React projects ⚛️
globs:
  - "**/*.jsx"
  - "**/*.tsx"
alwaysApply: false
---

- Use default React data binding to prevent XSS.
- Sanitize HTML before using dangerouslySetInnerHTML.
- Validate URLs before using them in the DOM.
- Secure server-side rendering and avoid JSON injection.
