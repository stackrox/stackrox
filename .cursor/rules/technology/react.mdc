# React-specific secure coding rules for Cursor
# This file provides security best practices for React projects.
---
description: Secure coding rules for React projects ⚛️
globs:
  - "**/*.jsx"
  - "**/*.tsx"
  - "**/src/**/*.jsx"
  - "**/src/**/*.tsx"
  - "**/components/**/*.jsx"
  - "**/components/**/*.tsx"
  - "**/pages/**/*.jsx"
  - "**/pages/**/*.tsx"
  - "**/App.jsx"
  - "**/App.tsx"
  - "**/index.jsx"
  - "**/index.tsx"
alwaysApply: false
---

- Use default React data binding to prevent XSS.
- Sanitize HTML before using dangerouslySetInnerHTML.
- Validate URLs before using them in the DOM.
- Secure server-side rendering and avoid JSON injection.
