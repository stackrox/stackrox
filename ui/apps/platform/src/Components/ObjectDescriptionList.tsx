import React, { ReactElement } from 'react';
import { DescriptionList, DescriptionListTerm } from '@patternfly/react-core';

import DescriptionListItem from 'Components/DescriptionListItem';

type LeafValue = number | number[] | string | string[] | null | undefined;

// Enough for PortConfig in Deployment, but intentionally not recursive.
type ObjectValue = Record<string, LeafValue | Record<string, LeafValue>[]>;

type ObjectDescriptionListProps = {
    data: ObjectValue;
    className?: string;
};

function ObjectDescriptionList({ data, className }: ObjectDescriptionListProps): ReactElement {
    return (
        <DescriptionList isHorizontal className={className}>
            {Object.entries(data).map(([key, value]) => {
                if (typeof value === 'number' || typeof value === 'string') {
                    return (
                        <div key={key}>
                            <DescriptionListItem term={key} desc={value} />
                        </div>
                    );
                }

                if (Array.isArray(value)) {
                    if (value.length === 0) {
                        return null;
                    }

                    const [item0] = value;
                    return (
                        <div key={key}>
                            {typeof item0 !== 'object' && (
                                <DescriptionListTerm>{key}</DescriptionListTerm>
                            )}
                            <ArrayDescriptionList data={value} className="pf-v5-u-pl-md" />
                        </div>
                    );
                }

                if (value === null || value === undefined) {
                    return null;
                }

                return (
                    <div key={key}>
                        <ObjectDescriptionList data={value} className="pf-v5-u-pl-md" />
                    </div>
                );
            })}
        </DescriptionList>
    );
}

type ArrayValue = number[] | string[] | Record<string, LeafValue>[];

type ArrayDescriptionListProps = {
    data: ArrayValue;
    className?: string;
};

function ArrayDescriptionList({ data, className }: ArrayDescriptionListProps): ReactElement {
    return (
        <DescriptionList isHorizontal className={className}>
            {data.map((value, key) => (
                // eslint-disable-next-line react/no-array-index-key
                <div key={key}>
                    {typeof value === 'number' || typeof value === 'string' ? (
                        <DescriptionListItem term={key} desc={value} />
                    ) : (
                        <ObjectDescriptionList data={value} className="pf-v5-u-pl-md" />
                    )}
                </div>
            ))}
        </DescriptionList>
    );
}

export default ObjectDescriptionList;
