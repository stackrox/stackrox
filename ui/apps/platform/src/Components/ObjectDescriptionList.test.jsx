import React from 'react';
import { render } from '@testing-library/react';

import ObjectDescriptionList from './ObjectDescriptionList';

const ports = [
    {
        name: '',
        containerPort: 22,
        protocol: 'TCP',
        exposure: 'UNSET',
        exposedPort: 0,
        exposureInfos: [],
    },
    {
        name: '',
        containerPort: 8080,
        protocol: 'TCP',
        exposure: 'INTERNAL',
        exposedPort: 0,
        exposureInfos: [
            {
                level: 'INTERNAL',
                serviceName: 'visa-processor-service',
                serviceId: 'eb1af0b8-a6da-11ea-a6c6-42010a800049',
                serviceClusterIp: '10.19.243.95',
                servicePort: 8080,
                nodePort: 0,
                externalIps: [],
                externalHostnames: [],
            },
        ],
    },
];

describe('ObjectDescriptionList', () => {
    test('annotations', () => {
        const { container } = render(
            <ObjectDescriptionList data={{ 'deprecated.daemonset.template.generation': '1' }} />
        );

        expect(container).toMatchSnapshot();
    });

    test('labels', () => {
        const { container } = render(<ObjectDescriptionList data={{ 'k8s-app': 'kube-proxy' }} />);

        expect(container).toMatchSnapshot();
    });

    test('ports', () => {
        const { container } = render(
            <>
                {ports.map((port, index) => (
                    // eslint-disable-next-line react/no-array-index-key
                    <ObjectDescriptionList key={index} data={port} />
                ))}
            </>
        );

        expect(container).toMatchSnapshot();
    });
});
