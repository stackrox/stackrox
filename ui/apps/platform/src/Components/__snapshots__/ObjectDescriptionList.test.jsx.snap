// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectDescriptionList annotations 1`] = `
<div>
  <dl
    class="pf-v5-c-description-list pf-m-horizontal"
  >
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            deprecated.daemonset.template.generation
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            1
          </div>
        </dd>
      </div>
    </div>
  </dl>
</div>
`;

exports[`ObjectDescriptionList labels 1`] = `
<div>
  <dl
    class="pf-v5-c-description-list pf-m-horizontal"
  >
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            k8s-app
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            kube-proxy
          </div>
        </dd>
      </div>
    </div>
  </dl>
</div>
`;

exports[`ObjectDescriptionList ports 1`] = `
<div>
  <dl
    class="pf-v5-c-description-list pf-m-horizontal"
  >
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            name
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          />
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            containerPort
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            22
          </div>
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            protocol
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            TCP
          </div>
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            exposure
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            UNSET
          </div>
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            exposedPort
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            0
          </div>
        </dd>
      </div>
    </div>
  </dl>
  <dl
    class="pf-v5-c-description-list pf-m-horizontal"
  >
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            name
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          />
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            containerPort
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            8080
          </div>
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            protocol
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            TCP
          </div>
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            exposure
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            INTERNAL
          </div>
        </dd>
      </div>
    </div>
    <div>
      <div
        class="pf-v5-c-description-list__group"
      >
        <dt
          class="pf-v5-c-description-list__term"
        >
          <span
            class="pf-v5-c-description-list__text"
          >
            exposedPort
          </span>
        </dt>
        <dd
          class="pf-v5-c-description-list__description"
        >
          <div
            class="pf-v5-c-description-list__text"
          >
            0
          </div>
        </dd>
      </div>
    </div>
    <div>
      <dl
        class="pf-v5-c-description-list pf-m-horizontal pf-v5-u-pl-md"
      >
        <div>
          <dl
            class="pf-v5-c-description-list pf-m-horizontal pf-v5-u-pl-md"
          >
            <div>
              <div
                class="pf-v5-c-description-list__group"
              >
                <dt
                  class="pf-v5-c-description-list__term"
                >
                  <span
                    class="pf-v5-c-description-list__text"
                  >
                    level
                  </span>
                </dt>
                <dd
                  class="pf-v5-c-description-list__description"
                >
                  <div
                    class="pf-v5-c-description-list__text"
                  >
                    INTERNAL
                  </div>
                </dd>
              </div>
            </div>
            <div>
              <div
                class="pf-v5-c-description-list__group"
              >
                <dt
                  class="pf-v5-c-description-list__term"
                >
                  <span
                    class="pf-v5-c-description-list__text"
                  >
                    serviceName
                  </span>
                </dt>
                <dd
                  class="pf-v5-c-description-list__description"
                >
                  <div
                    class="pf-v5-c-description-list__text"
                  >
                    visa-processor-service
                  </div>
                </dd>
              </div>
            </div>
            <div>
              <div
                class="pf-v5-c-description-list__group"
              >
                <dt
                  class="pf-v5-c-description-list__term"
                >
                  <span
                    class="pf-v5-c-description-list__text"
                  >
                    serviceId
                  </span>
                </dt>
                <dd
                  class="pf-v5-c-description-list__description"
                >
                  <div
                    class="pf-v5-c-description-list__text"
                  >
                    eb1af0b8-a6da-11ea-a6c6-42010a800049
                  </div>
                </dd>
              </div>
            </div>
            <div>
              <div
                class="pf-v5-c-description-list__group"
              >
                <dt
                  class="pf-v5-c-description-list__term"
                >
                  <span
                    class="pf-v5-c-description-list__text"
                  >
                    serviceClusterIp
                  </span>
                </dt>
                <dd
                  class="pf-v5-c-description-list__description"
                >
                  <div
                    class="pf-v5-c-description-list__text"
                  >
                    10.19.243.95
                  </div>
                </dd>
              </div>
            </div>
            <div>
              <div
                class="pf-v5-c-description-list__group"
              >
                <dt
                  class="pf-v5-c-description-list__term"
                >
                  <span
                    class="pf-v5-c-description-list__text"
                  >
                    servicePort
                  </span>
                </dt>
                <dd
                  class="pf-v5-c-description-list__description"
                >
                  <div
                    class="pf-v5-c-description-list__text"
                  >
                    8080
                  </div>
                </dd>
              </div>
            </div>
            <div>
              <div
                class="pf-v5-c-description-list__group"
              >
                <dt
                  class="pf-v5-c-description-list__term"
                >
                  <span
                    class="pf-v5-c-description-list__text"
                  >
                    nodePort
                  </span>
                </dt>
                <dd
                  class="pf-v5-c-description-list__description"
                >
                  <div
                    class="pf-v5-c-description-list__text"
                  >
                    0
                  </div>
                </dd>
              </div>
            </div>
          </dl>
        </div>
      </dl>
    </div>
  </dl>
</div>
`;
